widget/numerical_entry.go

```diff
@@ -72,12 +177,253 @@ func (e *NumericalEntry) Keyboard() mobile.KeyboardType {
+			}
+		}
+		if r == e.thouSep {
+			if i == 0 {
+				return errors.New("thousand separator cannot be the first character")
+			} else if i == 1 && runes[0] == e.minus {
+				return errors.New("thousand separator cannot be the first character after minus")
+			} else if runes[i-1] == e.radixSep {
+				return errors.New("thousand separator cannot be immediately after radix separator")
+			} else if runes[i-1] == e.thouSep {
+				return errors.New("thousand separator cannot be immediately after thousand separator")
+			}
+
+		}
+	}
+	return nil
+}
+
+// getRuneForLocale checks if a rune is valid for the entry,
+// and returns the correct rune for the locale.
+func (e *NumericalEntry) getRuneForLocale(r rune) (rune, bool) {
+	if unicode.IsDigit(r) {
+		return r, true
+	}
+
+	switch r {
+	case '-': // hyphen - minus
+		fallthrough
+	case 0x2212: //mathematical minus
+		if e.AllowNegative {
+			return e.minus, true
+		} else {
+			return 0, false
+		}
+	case '.': // full stop
+		fallthrough
+	case ',': // comma
+		if r == e.radixSep || r == e.thouSep {
+			return r, true
+		}
+	case ' ': // space
+		fallthrough
+	case 0xa0: // non-breaking space
+		if e.thouSep == ' ' || e.thouSep == 0xa0 {
+			return e.thouSep, true
+		}
+	case '\'': // single quote
+		fallthrough
+	case 0x2019: // right single quote mark
+		if e.thouSep == '\'' || e.thouSep == 0x2019 {
+			return e.thouSep, true
+		}
+	default:
+		if r == e.radixSep || r == e.thouSep {
+			return r, true
+		}
+	}
+	return 0, false
+}
+
+// getValidText filters the input text to allow only valid characters
+// based on the current locale and the entry's configuration.
+func (e *NumericalEntry) getValidText(curText, text string) string {
+	var s strings.Builder
+	for i, r := range text {
+		rn, ok := e.getRuneForLocale(r)
+		if !ok {
+			continue
+		}
+		if rn == e.minus {
+			if !e.AllowNegative || !(len(curText) == 0) || i != 0 {
+				continue
+			}
+		}
+		if rn == e.radixSep {
+			if !e.AllowFloat {
+				continue
+
+			}
+
+		}
+		s.WriteRune(rn)
+	}
+	return s.String()
+}
+
+// makeParsable prepares the text for parsing by removing thousand separators,
+// replacing the minus sign with a standard minus, and replacing the radix
+// separator with a period. It also validates the text before processing.
+func (e *NumericalEntry) makeParsable(text string) (string, error) {
+	err := e.ValidateText(text)
+	if err != nil {
+		return "", err
+	}
+	t := text
+	t = strings.ReplaceAll(text, string(e.thouSep), "")
+	t = strings.Replace(t, string(e.minus), "-", -1)
+	t = strings.Replace(t, string(e.radixSep), ".", -1)
+	return t, nil
+}
+
+// getLocaleRunes determines the minus sign, radix, and thousand separator
+// characters for a given locale by formatting a number and extracting
+// the relevant characters. It returns the minus sign, radix, and thousand
+// separator runes.
+func (e *NumericalEntry) getLocaleRunes(locale string) {
```

Comment:
This isn't a getter so needs a better name - it seems like like it is setup

Reply:
getLocaleRunes changed to setupRunes and called from setup setup was moved out of NewNumericalEntry and NewNumericalEntryWithData.